test/unit/hpa.c

```diff
@@ -452,6 +467,36 @@ TEST_BEGIN(test_defer_time) {
 }
 TEST_END
 
+TEST_BEGIN(test_purge_no_infinite_loop) {
```

Comment:
While the test is a very good reproduction of the bug this PR is fixing, I am thinking that we should make it a more general test like: 1. testing a bunch of option combinations within a certain range; 2. complicate the scenario a little bit so that purging with dirty pages existing can be tested stably as well. WDYT?

If @interwq is good with the current condition, we can merge the fix first and we can generalize the test later.

Reply:
Actually, I was hoping for feedback about this test :)

I am not sure how useful it is to have when bug is fixed, because it is carefully crafted for this particular bug. I searched and didn't find much regression test in the code base.

Primary goal was to have as small and at the same time clean demonstration that bug is in place and proposed patch indeed fixes it. So I intentionally didn't add any extra logic in the test.

That's being said, I am fine with removing or extend tests whatever consensus here will be.